src/selector_widget.cpp

```diff
@@ -0,0 +1,433 @@
+
+	const bool rtl = _current_text_dir == TD_RTL;
+
+	auto [first, last] = this->vscroll->GetVisibleRangeIterators(filtered_list);
+
+	for (auto it = first; it != last; ++it) {
+		const Rect ir = line.Shrink(WidgetDimensions::scaled.framerect);
+
+		if (!this->shown[*it]) {
+			DrawFrameRect(ir, COLOUR_BROWN, FR_LOWERED);
+		}
+		this->profile.DrawSection(this, *it, ir.Indent(WidgetDimensions::scaled.fullbevel.Horizontal(), rtl));
+
+		line = line.Translate(0, row_height);
+	}
+}
+
+/**
+ * This function is one of the possible functions ment to be used as part of the DrawWidget::Profile
+ * It draws "section" (line) of the scrollable list
+ * @param wid The current SelectionWidget object pointer
+ * @param id the "id" of the item to be drawn, can mean different things depending on the profile
+ * e.g when this widget displays companies, the id is a unique and valid CompanyID
+ * @param r the rectangle where the item is to be drawn
+ * @see SelectorWidget::DrawSection()
+ * @see SelectorWidget::DrawWidget()
+ * @see SelectorWidget::Profile
+ */
+void DrawSectionCompany(SelectorWidget *wid, int id, const Rect &r)
+{
+	const CompanyID cid = (CompanyID)id;
+	assert(Company::IsValidID(cid));
+
+	const bool rtl = _current_text_dir == TD_RTL;
+	const Dimension d = GetSpriteSize(SPR_COMPANY_ICON);
+
+	DrawCompanyIcon(cid, rtl ? r.right - d.width : r.left, CenterBounds(r.top, r.bottom, d.height));
+
+	const Rect text = r.Indent(d.width + WidgetDimensions::scaled.hsep_normal, rtl);
+
+	SetDParam(0, cid);
+	if (wid->selected_id.value_or(INVALID_OWNER) == cid) {
+		DrawString(text.left, text.right, CenterBounds(text.top, text.bottom, GetCharacterHeight(FS_NORMAL)), STR_COMPANY_NAME,  TC_WHITE);
+	} else {
+		DrawString(text.left, text.right, CenterBounds(text.top, text.bottom, GetCharacterHeight(FS_NORMAL)), STR_COMPANY_NAME,  TC_BLACK);
+	}
+}
+
+/**
+ * This function is one of the possible functions ment to be used as part of the DrawWidget::Profile
+ * It draws "section" (line) of the scrollable list
+ * @param wid The current SelectionWidget object pointer
+ * @param id the "id" of the item to be drawn, can mean different things depending on the profile
+ * e.g when this widget displays companies, the id is a unique and valid CompanyID
+ * @param r the rectangle where the item is to be drawn
+ * @see SelectorWidget::DrawSection()
+ * @see SelectorWidget::DrawWidget()
+ * @see SelectorWidget::Profile
+ */
+static void DrawSectionCargo(SelectorWidget *wid, int id, const Rect &r)
+{
+	const CargoID cargo_id = (CargoID)id; ///< Id of the current row's cargo
+	const CargoSpec *cargo = CargoSpec::Get(id); ///< cargo spec of the current row's cargo
+
+	const bool rtl = _current_text_dir == TD_RTL;
+
+	const float cargo_rect_scale_factor = 0.85f;
+	const int legend_width = r.Height() * 9 / 6;
+
+	/* Make a rectangle to display the legend color */
+	const Rect cargo_colour = r.Shrink(0, r.Height() * (1.0 - cargo_rect_scale_factor) / 2.0)
```

Comment:
`cargo_colour` is a confusing name, because it doesn't store the actual colour but a box that the colour is drawn in. Possible alternative:

```suggestion
	const Rect cargo_swatch = r.Shrink(0, r.Height() * (1.0 - cargo_rect_scale_factor) / 2.0)
```